### PR TITLE
[wip]fix(Core/Exploit)Summons that lead in to an instance are relocated when the instance is reset.

### LINKED ISSUE #10708: Instance Reset Port/Summon Exploit
https://github.com/chromiecraft/chromiecraft/issues/2983

### What client do you play on?

enUS

### Faction

- [X] Alliance
- [ ] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [ ] 60

### Current Behaviour

Entering an instance via summon takes you to the location you were summoned, this is easily exploited to quickly farm 5x lockouts of specific bosses.

Mostly an innocuous trick since nobody is harmed, and 5 lockouts isn't really the end of the world, but I felt it worth mentioning in case you felt it warranted fixing.

### Expected Blizzlike Behaviour

I can't find cases to support this being Blizzlike behaviour or not in a 3.3.5 setting, but I can say that in Vanilla and TBC if you want to quickly return to the start of an instance you can invite a party member outside the dungeon, everybody inside logs off, and the guy outside resets, but this sends you to the start of the instance, likely to prevent this very same sort of boss-farming. I would imagine post-"summoning inside instances" similar behaviour would have been implemented to summon you to the entrance, instead of where you left-off.

### Source

Here's the trick being using in DM:N, you could in theory farm DM:T chests with this, but it also works for any instance (besides resetting raid bosses, of course).

https://cdn.discordapp.com/attachments/938534369411358751/939955458280661062/Omegalul.webm

### Steps to reproduce the problem

Go to a boss inside an instance with a mage, warlock, and +1.
Kill boss
Make a summoning stone
Make a mage portal
Summon everybody
Everybody click the portal
Leader resets
Accept summon
Return to boss, but in a fresh instance.

### Extra Notes

_No response_

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/6f20b9ece71984f1a11469e65ede8accda05653e

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [mod-save-inventory](https://github.com/Nyeriah/mod-save-inventory)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by DepTypes**
## Changes Proposed:
-  Fixes the below mentioned exploit by removing summons in to instances that are reset.

## Issues Addressed:
- Closes #10708 

## SOURCE:
When resetting a dungeon for a player, additionally calls a new function:
`player->ClearSummonsToMapId(save->GetMapId());`

```
WorldPacket data(SMSG_SUMMON_CANCEL, 4);
data << uint32(mapid);
GetSession()->SendPacket(&data);
```
This WorldPacket properly signals to the client that the summon is no longer available, and will vanish immediately.

## Tests Performed:
Tested on Windows VS2019

2 Videos at the end.
Tested both resetting a portal in to and outside of an instance.
Behavior was summon would vanish when instance was reset, when a player was summoned inside the dungeon. 
When summoned outside of a dungeon, but dungeons are reset. No change happens.

## Known Issues and TODO List:

- [ ] The summon could be changed to teleport you to the door of the dungeon instead of being removed.
- [ ] I already wrote an example base for it. `Player::SetSummonToInstanceDoor(uint32 mapid)` but sadly the position I received from `MapEntry->GetEntrancePos(mapEntrance, x, y);` was wildly off. Needs another source for the coordinates.

https://user-images.githubusercontent.com/99797316/154498612-ceca5360-de9d-49d9-9e23-3108e5298239.mp4

https://user-images.githubusercontent.com/99797316/154498867-ec7b5f11-0ff7-4ba5-8f19-69c245fc5e0e.mp4

### CODE REVIEW

**Comment by DepTypes** [comment]
`Player::SetSummonToInstanceDoor(uint32)`

Which is spewing the warnings on gcc is not even part of code that is executed.
The reason I left it in was if someone comes up with an idea how to get the coordinates of the instance entrance properly.
This does not work as the coordinates are way off.

will fix the mad compilers later

**Comment by DepTypes** [comment]
Committed to fixing it 100% so that it changes the teleport location to the entrance if you old location was in an instance that got reset. Added settings to world.conf to instead send a direct cancellation of the summon.

Any ideas how to obtain accurate coordinates of the instance door? The MapManager is inaccurate at best, and doesn't provide height at all.

**Comment by DepTypes** [comment]
> If you are aiming for this to be a anticheat feature as per your conf there is a anticheat module u can migrate it to. image https://github.com/azerothcore/mod-anticheat

If you look at the original issue. It was explained how people can abuse this in dungeons by summoning everyone, taking a portal out, quickcly reset and be back at the last boss or whatever by taking the summon. The idea is to re-position the suommoned people to the entrance of the instance if someone does this, which I believe is valid behavior for the game.

Simply canceling the summon might hurt someone trying to just play with his friends.

**Comment by DepTypes** [comment]
Finding the instance door using
`const AreaTriggerTeleport* goBackTriggerTeleport = sObjectMgr->GetGoBackTrigger(mapid);`

Gets rid of unnecessary hacks.

**Comment by Lynethris** [comment]
Work but I got this error in worldserver :

Screenshot 2022-03-11 165124

**Comment by DepTypes** [comment]
> Work but I got this error in worldserver :
> 
> Screenshot 2022-03-11 165124

Does this error happen when resetting instance with any summoned GameObject inside instance (mage table)? I have to test later, but from only that error message it seems a bit out of scope for this PR.

TODO:

- Test with SpellId 61993
- Test resetting instances with other summoned game objects inside instance
- If the bug happens with other GOs, consider opening new issue...

File: src/server/game/Instances/InstanceSaveMgr.cpp [hunk 1]
```diff
@@ -789,5 +789,11 @@ void InstanceSaveMgr::UnbindAllFor(InstanceSave* save)
 {
     GuidList& pList = save->m_playerList;
     while (!pList.empty())
```
Comment (DepTypes) [inline src/server/game/Instances/InstanceSaveMgr.cpp:791]: This needs to be rewritten. For ref: https://github.com/azerothcore/azerothcore-wotlk/pull/11003/

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -1321,6 +1321,65 @@ void Player::SendTeleportAckPacket()
...
+// Used to notify client, their summon has been invalidated
+void Player::ExpireSummonTimer(int32 mapid = -1)
+{
+    // Set server summon timer to expired
+    this->m_summon_expire = 0;
+
+    // If no mapid passed, find it
+    // Mapid is possibly used to differentiate between concurrent summons
+    if (mapid == -1)
+        mapid = this->m_summon_mapid;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Player/Player.cpp:1333]: brackets

**Comment by Nyeriah** [comment]
No response from the author after two weeks, also unresolved change requests dating prior to that. Marking as abandoned. If the author decides to give this another go, just drop a message and we will reopen it.

**Pull request closed without merging**